torch/csrc/jit/c10_ops/layer_norm.cpp

```diff
@@ -0,0 +1,33 @@
+#include <c10/core/dispatch/Dispatcher.h>
+#include <c10/core/opschema/layer_norm.h>
+#include <torch/csrc/jit/custom_operator.h>
+#include <torch/csrc/autograd/variable.h>
+#include <caffe2/core/context.h>
+
+using c10::C10Tensor;
+
+namespace {
+// TODO Return tuple<Tensor, Tensor, Tensor> instead of vector<Tensor>
+std::vector<at::Tensor> layer_norm(
+    at::Tensor input,
+    int64_t axis,
+    double epsilon) {
+  if (input.requires_grad()) {
+    throw std::runtime_error("Autograd not yet supported for c10 ops.");
+  }
+  c10::core::opschema::LayerNorm::Cache cache;
+  C10Tensor c10_input(torch::autograd::Variable(std::move(input)).data());
+  C10Tensor c10_output(at::empty({0}));
+  C10Tensor c10_output_mean(at::empty({0}));
+  C10Tensor c10_output_stdev(at::empty({0}));
+  c10::Dispatcher<c10::core::opschema::LayerNorm>::call(c10_input, c10_output, c10_output_mean, c10_output_stdev, (int)axis, (float)epsilon, &cache);
+  return {
+    torch::autograd::make_variable(at::Tensor(std::move(c10_output)), false),
```

Comment:
Yeah, this is wrong. If you do this, you'll generate a leaf variable that doesn't require grad. This will lead to hard to debug problems if someone tries to differentiate through this, because the gradients will just stop at the outputs of this op and accumulate there.

Reply:
Also CC @goldsborough because I know autograd support for custom operators is not done yet, and I don't know what the eventual design was planned to be.

Reply:
Afaik, this will always get a variable and should always return a variable. However, this variable might have `requires_grad==true` or `requires_grad==false`. This code only handles `requires_grad=false` correctly, that's why there is an exception thrown in line 16 if `requires_grad==true`.

Reply:
OK... that seems convincing enough.

Still, all of this boilerplate seems like a massive antipattern. The way you have structured the code here, seems to imply that everyone who wants to write an operator this way has to write the boilerplate of unwrapping and then rewrapping the variable. This is why we had code generation in the old version of this code. If you are going to say that, "Ah, I'll figure out how to solve this problem later," it would be nice if this code, somewhere, acknowledged the issue, rather than having no comments at all. But I am not sure at all how you are going to solve this problem; we talked about a boxed representation so that we can uniformly perform common operator operations, including wrapping and unwrapping, but I am afraid that it has passed out of the collective memories of everyone. Doc is here https://fb.quip.com/o5pXA1LHgAq0 or in this Gist for OSS people following: https://gist.github.com/ezyang/0ff02f92d742c2587fd76b640eef9e11 )

Also cc @yf225 because all of this changes when Tensor and Variable merge in C++ (but this doesn't "solve" your problem, because you still need to deal with requires grad = true).

Reply:
I added a detailed comment.